client-mr/src/main/java/org/apache/hadoop/mapred/SortWriteBuffer.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Tencent is pleased to support the open source community by making
+ * Firestorm-Spark remote shuffle server available. 
+ *
+ * Copyright (C) 2021 THL A29 Limited, a Tencent company.  All rights reserved. 
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * https://opensource.org/licenses/Apache-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package org.apache.hadoop.mapred;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Comparator;
+import java.util.List;
+
+import com.google.common.collect.Lists;
+import org.apache.hadoop.io.RawComparator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SortWriteBuffer<K, V> extends OutputStream  {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SortWriteBuffer.class);
+  private long copyTime = 0;
+  private final List<WrappedBuffer> buffers = Lists.newArrayList();
+  private final List<Record<K>> records = Lists.newArrayList();
```

Comment:
one record will be stored in both buffers(for serialized) and records(for order)?

Reply:
No, we only store keys in the records.

Reply:
Data size in records is ignored, and it will cause memory problem.
For example, in "word count" case, data size of key can't be ignored.

Reply:
> Data size in records is ignored, and it will cause memory problem. For example, in "word count" case, data size of key can't be ignored.
Let me think of another method.

Reply:
> > Data size in records is ignored, and it will cause memory problem. For example, in "word count" case, data size of key can't be ignored.
> > 
Let me think of another method.

Reply:
It seems two methods to solve this,  We don't serialize key at fist, when we sort finish, we serialize the keys.But it will be too difficult.I think it will be more easy if we sort the block  in the reduce side.

Reply:
> 

If sort is happened only in reduce task, can't reuse sort implementation from MR.
For this problem, we need get a better solution first.

Reply:
modify as we discussed.